test/gpu/native/studies/shoc/sort.chpl

```diff
@@ -0,0 +1,679 @@
+        var sStartPointers = startVoidPtr : c_ptr(uint(32));
+
+        var sRadixVoidPtr = __primitive("gpu allocShared", 2*SCAN_BLOCK_SIZE*numBytes(uint(32)));
+        var sRadix1 = sRadixVoidPtr : c_ptr(uint(32));
+
+        // var groupId = __primitive("gpu blockIdx x") :uint(32);
+        const groupId = i / SCAN_BLOCK_SIZE : uint(32);
+        // var localId = __primitive("gpu threadIdx x"): uint(32);
+        const localId = i % SCAN_BLOCK_SIZE: uint(32);
+        const groupSize = SCAN_BLOCK_SIZE : uint(32);
+
+        var radix2 : 2*uint(32);
+        radix2[0] = keys[2 * i];
+        radix2[1] = keys[2 * i + 1];
+
+        sRadix1[2 * localId]     = (radix2[0] >> startbit) & 0xF;
+        sRadix1[2 * localId + 1] = (radix2[1] >> startbit) & 0xF;
+
+        // Finds the position where the sRadix1 entries differ and stores start
+        // index for each radix.
+        if localId < 16 then sStartPointers[localId] = 0;
+
+        __primitive("gpu syncThreads");
+
+        if (localId > 0) && (sRadix1[localId] != sRadix1[localId - 1]) then
+            sStartPointers[sRadix1[localId]] = localId;
+
+        if sRadix1[localId + groupSize] != sRadix1[localId + groupSize - 1] then
+            sStartPointers[sRadix1[localId + groupSize]] = localId + groupSize;
+
+        __primitive("gpu syncThreads");
+
+        if localId < 16 then
+            blockOffsets[groupId*16 + localId] = sStartPointers[localId];
+
+        __primitive("gpu syncThreads");
+
+        // Compute the size of each block
+        if (localId > 0) && (sRadix1[localId] != sRadix1[localId - 1]) then
+            sStartPointers[sRadix1[localId - 1]] =
+                localId - sStartPointers[sRadix1[localId - 1]];
+
+        if sRadix1[localId + groupSize] != sRadix1[localId + groupSize - 1] then
+          sStartPointers[sRadix1[localId + groupSize - 1]] =
+               localId + groupSize - sStartPointers[sRadix1[localId +
+                                                            groupSize - 1]];
+
+
+        if localId == groupSize - 1 then
+            sStartPointers[sRadix1[2 * groupSize - 1]] =
+                2 * groupSize - sStartPointers[sRadix1[2 * groupSize - 1]];
+
+        __primitive("gpu syncThreads");
+
+        if localId < 16 then
+            counters[localId * totalBlocks + groupId] = sStartPointers[localId];
+    }
+
+}
+
+proc reorderData (reorderGlobalWorkSize, startbit: uint(32),
+        ref outKeys : [] uint(32), ref outValues : [] uint(32),
+        ref keys : [] uint(32), ref values : [] uint(32),
+        ref blockOffsets : [] uint(32), ref offsets : [] uint(32),
+        ref sizes : [] uint(32), totalBlocks : uint(32)) {
+
+    forall i in 0..<reorderGlobalWorkSize: uint(32){
+        __primitive("gpu set blockSize", SCAN_BLOCK_SIZE);
+
+        const GROUP_SIZE = SCAN_BLOCK_SIZE : uint(32);
+
+        var sKeys2VoidPtr = __primitive("gpu allocShared", 512*numBytes(uint(32)));
```

Comment:
I'd add a comment to this (and sValues2VoidPtr) and any other variables that this was a `uint2` type in the original CUDA code but you're flattening it into an array (so you're allocating 512 elements instead of the 256 in the original code)